src/main/java/org/sonarsource/sonarlint/ls/AnalysisManager.java

```diff
@@ -0,0 +1,452 @@
+  }
+
+  static Optional<Diagnostic> convert(Issue issue) {
+    if (issue.getStartLine() != null) {
+      Range range = position(issue);
+      Diagnostic diagnostic = new Diagnostic();
+      DiagnosticSeverity severity = severity(issue.getSeverity());
+
+      diagnostic.setSeverity(severity);
+      diagnostic.setRange(range);
+      diagnostic.setCode(issue.getRuleKey());
+      diagnostic.setMessage(issue.getMessage());
+      diagnostic.setSource(SONARLINT_SOURCE);
+
+      List<Flow> flows = issue.flows();
+      // If multiple flows with more than 1 location, keep only the first flow
+      if (flows.size() > 1 && flows.stream().anyMatch(f -> f.locations().size() > 1)) {
+        flows = Collections.singletonList(flows.get(0));
+      }
+      diagnostic.setRelatedInformation(flows
+        .stream()
+        .flatMap(f -> f.locations().stream())
+        // Message is mandatory in lsp
+        .filter(l -> nonNull(l.getMessage()))
+        // Ignore global issue locations
+        .filter(l -> nonNull(l.getInputFile()))
+        .map(l -> {
+          DiagnosticRelatedInformation rel = new DiagnosticRelatedInformation();
+          rel.setMessage(l.getMessage());
+          rel.setLocation(new Location(l.getInputFile().uri().toString(), position(l)));
+          return rel;
+        }).collect(Collectors.toList()));
+
+      return Optional.of(diagnostic);
+    }
+    return Optional.empty();
+  }
+
+  private static DiagnosticSeverity severity(String severity) {
+    switch (severity.toUpperCase(Locale.ENGLISH)) {
+      case "BLOCKER":
+      case "CRITICAL":
+        return DiagnosticSeverity.Error;
+      case "MAJOR":
+        return DiagnosticSeverity.Warning;
+      case "MINOR":
+        return DiagnosticSeverity.Information;
+      case "INFO":
+      default:
+        return DiagnosticSeverity.Hint;
+    }
+  }
+
+  private static Range position(Issue issue) {
+    return new Range(
+      new Position(
+        issue.getStartLine() - 1,
+        issue.getStartLineOffset()),
+      new Position(
+        issue.getEndLine() - 1,
+        issue.getEndLineOffset()));
+  }
+
+  private static Range position(IssueLocation location) {
+    return new Range(
+      new Position(
+        location.getStartLine() - 1,
+        location.getStartLineOffset()),
+      new Position(
+        location.getEndLine() - 1,
+        location.getEndLineOffset()));
+  }
+
+  private static PublishDiagnosticsParams newPublishDiagnostics(URI newUri) {
+    PublishDiagnosticsParams p = new PublishDiagnosticsParams();
+
+    p.setDiagnostics(new ArrayList<>());
+    p.setUri(newUri.toString());
+
+    return p;
+  }
+
+  public void initialize(@Nullable Path typeScriptPath) {
+    this.typeScriptPath = typeScriptPath;
+    watcher.start();
+  }
+
+  public void shutdown() {
+    watcher.stopWatcher();
+    eventMap.clear();
+    analysisExecutor.shutdown();
+    if (standaloneEngine != null) {
+      standaloneEngine.stop();
+    }
```

Comment:
Would it make sense to call `watcher.join()` here?

Reply:
I don't think we need to. This is a daemon thread, so it won't prevent JVM to exit.